### ISSUE TITLE
Implement the abstract class for cameras

### ISSUE THREAD

**Issue by edyoshikun** [czbiohub/coPylot#166]
There is a need to define what the abstract classes need for the cameras. This will help develop us write modular code.  Taking inspiration from micromanager [here](https://github.com/micro-manager/mmCoreAndDevices/blob/4a9828bfb73d073cb37b70297f467721551da08d/MMDevice/MMDevice.h#L345) I believe the core properties to implement are the following:
- Snap()
- Get/Set Exposure()
- Get/Set Framerate()
- Get/Set Image size() 
- Get/Set Gain()
- Get/Set BitDepth()
- Get/Set Binning()
- Get/Set Trigger()
- Get Min/Max Exposure, Gain, Sensor Size

Please feel free to add more. @AhmetCanSolak @ieivanov @YangLiujames

**Comment by YangLiujames** [comment]
Camera we are dealing with often have different running mode.. typical light sheet mode or normal rolling shutter mode.. area mode...and the amount delay for each line...

**Comment by edyoshikun** [comment]
> Get/Set Framerate()

Yes, most cameras have a fixed framerate, but the FLIR and basler have settings for you to change that. 
By default and the most used scenario is that the frame rate is always at the highest possible and limited by the exposure time+ readout. I don't really use this , but since the abstract class is to cover most cameras perhaps someone has a use case(?) I am happy if this function does not exist.

> Get/Set Trigger()

Trigger type maybe was a very broad term. In my experience setting triggers involves mainly two settings:
1) rising/falling edge and if it can overlap
2) Triggering single/burst frame

+1 `shutter_mode`
+1 `list_available_cameras`

One property I didn't add is:
`Get/Set Readout_mode()` I know some cameras like the Andor and the Hamamatsu have the choice of different clock-speeds for readout. I am not aware of the FLIR or Basler having something like that, but I haven't looked closely.

**Comment by YangLiujames** [comment]
The shutter mode and readout time are somewhat interconnected.  I am in favor of keep a Get Framerate() but not Set...since most of the sCMOS don't have that option. These two I found pretty useful as well.

1. Triggertype() selector
2.  Output signal(s)

**Comment by AhmetCanSolak** [comment]
could you explain what you refer as output signals more @YangLiujames ?

**Comment by YangLiujames** [comment]
1. Tirggertype -- this for external trigger , such as  Edge trigger / Global reset edge trigger , Level trigger / Global reset level trigger, Synchronous readout trigger, Start trigger (ORCA flash as an example , see page 48) https://camera.hamamatsu.com/content/dam/hamamatsu-photonics/sites/static/sys/en/manual/C11440-42U30_IM_En.pdf
2. Output signal(s)
This select the kind of output that camera send. if there are multiple output ports, we can config each output to different alliable options, such as , fire all, fire any or read out start/end , page 36
(https://andor.oxinst.com/downloads/uploads/Zyla_hardware_user_guide.pdf)
